lib/smack/SmackInstGenerator.cpp

```diff
@@ -889,28 +889,133 @@ void SmackInstGenerator::visitIntrinsicInst(llvm::IntrinsicInst& ii) {
+    unsigned offset = width&1;
+    for (unsigned i = width%2==0 ? 1 : 0; i < width/2; ++i) {
+      body = Expr::bvConcat(Expr::bvConcat(Expr::bvExtract(var,width/2-i,width/2-i-1),
+                                           body),
+                            Expr::bvExtract(var, width/2+i+1+offset, width/2+i+offset));
+    }
+    return body;
+  };
+
+  static const auto bswap = [this](Value* arg){
+    auto width = arg->getType()->getIntegerBitWidth();
+    auto var = rep->expr(arg);
+
+    // Swap the bytes to the right and left of the middle
+    const Expr* body = Expr::bvConcat(Expr::bvExtract(var,width/2,width/2-8),
+                                      Expr::bvExtract(var,width/2+8,width/2));
 
-  static const auto bswap = [this](CallInst* ci){
+    // Swap the bytes to the right and the left of the already swapped portion.
+    for (unsigned i = 8; i < width/2; i += 8) {
+      body = Expr::bvConcat(Expr::bvConcat(Expr::bvExtract(var,width/2-i,width/2-i-8),
+                                           body),
+                            Expr::bvExtract(var, width/2+i+8, width/2+i));
+    }
+    return body;
+  };
+
+  // Count leading zeros
+  static const auto ctlz = [this](CallInst* ci) {
     if (SmackOptions::BitPrecise) {
       auto width = ci->getArgOperand(0)->getType()->getIntegerBitWidth();
       auto var = rep->expr(ci->getArgOperand(0));
 
-      // Swap the bytes to the right and left of the middle
-      const Expr* body = Expr::bvConcat(Expr::bvExtract(var,width/2,width/2-8),
-                                        Expr::bvExtract(var,width/2+8,width/2));
+      // e.g., if v[32:31] == 1 then 0bv32 else if v[31:30] == 1 then 1bv32 else
+      // ... else if v[1:0] == 1 then 31bv32 else 32bv32
+      const Expr* body = Expr::lit(width, width);
+      for (unsigned i = 0; i < width; ++i) {
+        body = Expr::if_then_else(Expr::eq(Expr::bvExtract(var, i+1, i),
+                                           Expr::lit(1,1)),
+                                  Expr::lit(width-i-1, width),
+                                  body);
+      }
 
-      // Swap the bytes to the right and the left of the already swapped portion.
-      for (unsigned i = 8; i < width/2; i += 8) {
-        body = Expr::bvConcat(Expr::bvConcat(Expr::bvExtract(var,width/2-i,width/2-i-8),
-                                             body),
-                              Expr::bvExtract(var, width/2+i+8, width/2+i));
+      // Handle the is_zero_undef case, i.e. if the flag is set and the argument
+      // is zero, then the result is undefined.
+      auto isZeroUndef = rep->expr(ci->getArgOperand(1));
+      body = Expr::if_then_else(Expr::and_(Expr::eq(isZeroUndef, Expr::lit(1, 1)),
+                                           Expr::eq(var, Expr::lit(0,width))),
+                                rep->expr(ci), // The result is undefined
+                                body);
+      emit(Stmt::havoc(rep->expr(ci)));
```

Comment:
Why do you have this havoc before the assignment?

Reply:
Basically what this is doing is this:
```
$i4 = if (($is_zero_undef == 1bv1) &&
              ($i1 == 0bv32))
          then $i4 else
          ctlz($i4)
```
The havoc is there to make `$i4` nondeterministic since LLVM says this is undefined. This comes up if this is being done in a loop where `$i4` gets an old value. I added an example which passes if it's not havoc'd first.

Perhaps it would be better to have an assertion failure instead?

Reply:
Related to #406. The `is_zero_undef` argument to `llvm.ctz.*(%arg, %is_zero_undef)` is a constant, so we could emit a warning if `is_zero_undef==true` that the user may want to use another flag (something like `--undefined`) if they get weird results.

Reply:
Ah, ok, I get it. What do you mean by "to have an assertion failure instead"?
I also don't quite get your last comment.

Reply:
What I'm thinking of is that someone gets an assertion violation from code using this function with zero undefined, the cause may not be very obvious. I'm thinking we can give a simple warning that this can happen and use a general undefined behavior assertion category for catching these errors. As it is, I think this is well outside the scope of this PR.